### DISCUSSION TITLE
Efficient implementation of nonlinear dynamics and constraints

### DISCUSSION THREAD

**Discussion by afossa** [infiniteopt/InfiniteOpt.jl#371] [Q&A]
Hello,

I am using `InfiniteOpt` to solve optimal control problems in trajectory optimization. These type of problems frequently have complicated nonlinear expressions in both the dynamics and the constraints, and they rely a lot on operations like matrix-vector products, cross products, norms, etc.

I was able to successfully solve an optimal transfer between the Earth and Mars, but I am looking for suggestions on how I can improve my implementation, specifically how to avoid recomputing quantities that appear in more than one constraint, such as the distance from the Sun. Should I add a variable for each of them, or is there a better way to achieve this?

I am also experimenting with different problem formulations, in particular I want to express the control variables for the three components of the thrust vector in either the inertial frame or in a rotating frame defined by the spacecraft position and velocity vectors. In the second case, I need to compute the rotation matrix from the rotating to the inertial frame, but I am getting an error that seems to be related to the normalization of the angular momentum vector (computed as the cross product between position and velocity):

```julia
Unable to create a nonlinear expression because it did not contain any JuMP scalars. head = `:+`, args = `Any[0.0]`.
```

So I modified the problem formulation to avoid it, however, the accuracy of the solution with the thrust expressed in rotating frame is much lower than that of the solution with the thrust expressed directly in inertial frame. My guess is that the rotation matrix, as I am computing it now, is not perfectly orthogonal. This is the working code I came up with:

```julia
@variables(mdl, begin
    r[1:3], Infinite(t) # position vector
    v[1:3], Infinite(t) # velocity vector
    m_dry <= m <= m0, Infinite(t) # mass
    0.0 <= Γ, Infinite(t), (start=0.0) # energy integral
    0.0 <= T <= thr_max, Infinite(t), (start=0.0) # thrust magnitude
    u[1:3], Infinite(t), (start=0.0) # thrust components
end);

# zeroth-order hold control
constant_over_collocation(T, t);
[constant_over_collocation(u[i], t) for i in 1:3];

# initial guess
[set_start_value_function(r[i], t -> linear_interpolation(tv_l, rvv_l[i,:], extrapolation_bc=Line())(t)) for i in 1:3];
[set_start_value_function(v[i], t -> linear_interpolation(tv_l, rvv_l[i+3,:], extrapolation_bc=Line())(t)) for i in 1:3];
set_start_value_function(m, t -> linear_interpolation([t0, tf], [m0, m_dry], extrapolation_bc=Line())(t));

if frm_id != 0 # thrust components in local orbital frame

    # unit vectors along position, velocity, and angular momentum vectors
    @expression(mdl, ur, r ./ dot(r,r)^0.5);
    @expression(mdl, uv, v ./ dot(v,v)^0.5);
    @expression(mdl, uh, [ur[2]*uv[3] - ur[3]*uv[2]; ur[3]*uv[1] - ur[1]*uv[3]; ur[1]*uv[2] - ur[2]*uv[1]]);

    # unit vectors defining the local orbital frame
    if frm_id == 1 # qsw frame
        @expression(mdl, ux, ur);
    else # tnw frame
        @expression(mdl, ux, uv);
    end;
    @expression(mdl, uz, uh);
    @expression(mdl, uy, [uz[2]*ux[3] - uz[3]*ux[2]; uz[3]*ux[1] - uz[1]*ux[3]; uz[1]*ux[2] - uz[2]*ux[1]]);

    # thrust components in inertial frame
    @expression(mdl, ui, [ux[1]*u[1] + uy[1]*u[2] + uz[1]*u[3]; ux[2]*u[1] + uy[2]*u[2] + uz[2]*u[3]; ux[3]*u[1] + uy[3]*u[2] + uz[3]*u[3]]);

    # thrust acceleration in inertial frame
    @expression(mdl, at, ui ./ m);
else # thrust components in inertial frame
    @expression(mdl, at, u ./ m); # thrust acceleration in inertial frame
end;

# gravitational acceleration in inertial frame
@expression(mdl, ag, -gm / dot(r,r)^1.5 * r);

# dynamics
@constraint(mdl, [i=1:3], ∂(r[i], t) == v[i]);
@constraint(mdl, [i=1:3], ∂(v[i], t) == ag[i] + at[i]);
@constraint(mdl, ∂(m, t) == -T / vex);
@constraint(mdl, ∂(Γ, t) == 0.5 * T^2);

# control constraints
@constraint(mdl, dot(u,u) <= T^2); # thrust magnitude
```

Do you have any suggestion on how I can improve it, both in terms of accuracy and efficiency as discussed above?

The results I obtain in inertial frame are:
```
Problem type:   energy optimal
Thrust vector:  inertial frame
Remaining fuel: 56.51349789895097 kg
Errors between final BCs and explicit simulation:
Position: 1.667613622724852 km
Velocity: 6.277496265271775e-8 km/s
```
and in QSW frame:
```
Problem type:   energy optimal
Thrust vector:  qsw frame
Remaining fuel: 55.46058819518962 kg
Errors between final BCs and explicit simulation:
Position: 903149.9219073891 km
Velocity: 0.12782349401040421 km/s
```
finally, in TNW frame:
```
Problem type:   energy optimal
Thrust vector:  tnw frame
Remaining fuel: 54.81937906234515 kg
Errors between final BCs and explicit simulation:
Position: 492797.71709358005 km
Velocity: 0.09807832598412268 km/s
```

Attached is the full script and input data to reproduce these results. I am using the version of `InfiniteOpt` checked out from the `master` branch to use the `constant_over_collocation` method that enforces a piecewise-constant control.

[earth_mars_input.json](https://github.com/user-attachments/files/18138533/earth_mars_input.json)
[earth_mars_script.txt](https://github.com/user-attachments/files/18138534/earth_mars_script.txt)

**Comment by pulsipher** [comment]
Hi @afossa,

I somehow missed this post, I am very sorry for the late reply. To better help you, could you please provide the complete script? From the above it is not clear how the infinite parameter is being defined.

**Reply by afossa** [reply]
Hello @pulsipher,

Here is the last iteration of the complete script, as well as the JSON file with all input data. Just rename the `.txt` into `.jl`.

[earth_mars.txt](https://github.com/user-attachments/files/19272793/earth_mars.txt)

[inputs.json](https://github.com/user-attachments/files/19272791/inputs.json)

I am running it with Julia `v1.11.4` and with `InfiniteOpt.jl` at the latest commit on `master`. To run the different simulations, just change the value of `frm_id` on line 13 to either `0`, `1`, or `2`.

The results I obtain on my side are the followings:

1. Thrust in inertial frame, i.e. `frm_id = 0`:
```
Problem type:   energy optimal
Thrust vector:  inertial frame
Remaining fuel: 56.51830328351519 kg

Errors between final BCs and explicit simulation:
Position: 9347.155667111127 km
Velocity: 0.0013605904319164755 km/s
```

2. Thrust in QSW frame, i.e. `frm_id = 1`:
```
Problem type:   energy optimal
Thrust vector:  qsw frame
Remaining fuel: 55.47266773857979 kg

Errors between final BCs and explicit simulation:
Position: 909190.0672712309 km
Velocity: 0.12884178326905255 km/s
```

3. Thrust in TNW frame, i.e. `frm_id = 2`:
```
Problem type:   energy optimal
Thrust vector:  tnw frame
Remaining fuel: 54.83086303420842 kg

Errors between final BCs and explicit simulation:
Position: 486970.22034870664 km
Velocity: 0.09693190054817039 km/s
```

**Reply by pulsipher** [reply]
Hi @afossa, this is a cool model and there is a lot to unpack. 

One problem likely at play here is the equation scaling. It seems that there are some very large difference in the values of the variables involved in different operations which it often problematic for numerical solvers to handle. I would suggest scaling the variables/equations such the orders-of-magnitude are similar.

**Reply by afossa** [reply]
Hi @pulsipher,

The optimization variables are already scaled in the problem formulation, and are all of order one in the `InfiniteOpt` model. I am only multiplying them back by their scaling factors after solving the problem, and only for output and visualization purposes.

You can check the range of values for the optimization variables by looking at the JSON file I attached above. After solving the scaled problem, you can recover their values in metric units by multiplying the outputs of the optimization by the values in the `scale` field.

**Reply by pulsipher** [reply]
What is the scaling factor for the position vector? A key thing to keep in mind is that the default tolerance for the Ipopt solver is 1e-8, whereas your integration is using a tolerance of 1e-14. Also, `Vern9` is a 9th order numerical scheme which is likely to be more accurate than the collocation scheme depending on the number of nodes that you choose.

**Reply by afossa** [reply]
The components of the positions are scaled by 1 AU, such that the initial radius is approximately 1 (the Earth's orbit radius in AU) and the final radius is approximately 1.5 (the Mars' orbit radius in AU). Velocities are scaled such that the period of the Earth orbit is 2pi.

For the discretization I am using 400 nodes and a third-order orthogonal collocation. I am also setting these tolerances for IPOPT
```
set_optimizer_attribute(mdl, "tol", 1e-14);
set_optimizer_attribute(mdl, "acceptable_tol", 1e-12);
```

And this is the output I obtain for the QSW case:
```

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit https://github.com/coin-or/Ipopt
******************************************************************************

This is Ipopt version 3.14.17, running with linear solver ma57.

Number of nonzeros in equality constraint Jacobian...:    62457
Number of nonzeros in inequality constraint Jacobian.:     3204
Number of nonzeros in Lagrangian Hessian.............:   136170

Total number of variables............................:    16020
                     variables with only lower bounds:      801
                variables with lower and upper bounds:     1602
                     variables with only upper bounds:        0
Total number of equality constraints.................:    14422
Total number of inequality constraints...............:      801
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:      801

iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  9.9999900e-03 2.21e+00 1.00e+00  -1.0 0.00e+00    -  0.00e+00 0.00e+00   0
Reallocating memory for MA57: lfact (3133656)
Reallocating memory for MA57: lfact (8710501)
Reallocating memory for MA57: lfact (9236184)
  10  1.6837663e-01 2.75e+00 2.23e+04  -1.0 1.55e+00   3.4 8.53e-01 4.24e-02h  1
  20  3.4073975e-01 1.65e+00 2.94e+06  -1.0 7.04e-01   5.9 1.00e+00 1.15e-01h  1
  30  4.6132396e-01 7.11e-01 8.32e+05  -1.0 4.09e-01   5.6 1.39e-01 1.30e-01h  1
  40  4.9038565e-01 4.69e-01 1.69e+06  -1.0 2.80e-01   6.3 9.82e-01 2.32e-02h  1
  50  5.1060617e-01 3.10e-01 1.04e+06  -1.0 1.96e-01   6.0 1.81e-01 2.26e-02h  1
  60  5.3107368e-01 1.54e-01 1.36e+05  -1.0 4.63e-01   4.9 2.21e-01 3.09e-02h  1
  70  5.4816519e-01 4.03e-02 4.42e+05  -1.0 3.31e-02   6.1 1.00e+00 5.45e-02h  1
  80  5.6205967e-01 7.27e-06 6.34e+02  -1.0 2.59e-03   4.5 1.00e+00 1.00e+00f  1
  90  5.5816608e-01 5.12e-03 1.94e+04  -1.7 1.54e-01    -  4.89e-01 1.00e+00f  1
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
 100  5.6610257e-01 8.92e-07 2.04e+00  -1.7 3.71e-03   2.8 1.00e+00 1.00e+00f  1
 110  6.1745954e-01 2.33e-06 8.93e-02  -2.5 1.35e-02   0.8 1.00e+00 1.00e+00h  1
 120  5.5042490e-01 6.09e-06 1.33e-03  -5.7 5.45e-03  -0.6 1.00e+00 1.00e+00h  1
 130  4.9443873e-01 2.63e-02 9.64e-05  -5.7 1.65e-01    -  1.00e+00 1.00e+00h  1
 140  4.9300224e-01 2.77e-08 7.72e-05 -12.9 3.58e-04    -  1.00e+00 1.00e+00h  1

Number of Iterations....: 144

                                   (scaled)                 (unscaled)
Objective...............:   4.9300224388589847e-01    4.9300224388589847e-01
Dual infeasibility......:   7.5414183501401621e-16    7.5414183501401621e-16
Constraint violation....:   8.5659322149167936e-16    8.5659322149167936e-16
Variable bound violation:   9.9989529989130688e-09    9.9989529989130688e-09
Complementarity.........:   9.1224240377574504e-16    9.1224240377574504e-16
Overall NLP error.......:   9.1224240377574504e-16    9.1224240377574504e-16


Number of objective function evaluations             = 145
Number of objective gradient evaluations             = 145
Number of equality constraint evaluations            = 145
Number of inequality constraint evaluations          = 145
Number of equality constraint Jacobian evaluations   = 145
Number of inequality constraint Jacobian evaluations = 145
Number of Lagrangian Hessian evaluations             = 144
Total seconds in IPOPT (w/o function evaluations)    =     11.486
Total seconds in NLP function evaluations            =      4.377

EXIT: Optimal Solution Found.
```

the outputs for the other two simulations are very similar. Finally, the explicit integration uses a piecewise-constant control that matches the solution to the NLP.

**Reply by pulsipher** [reply]
Given the relative scales involved. The discrepancy between the optimization solution and the integrator could simply be explained by the relative order of the methods used for integration (9th order instead of 3rd order). In other words, I suspect the error can just be attributed to truncation error. The formulation itself looks reasonable. A tolerance of `1e-14` is also on the small side for operations with `Float64`, unfortunately most NLP solvers like Ipopt do not support `BigFloat`.
